### DISCUSSION TITLE
Deserialize OffsetDateTime from serde_json::Value

### DISCUSSION THREAD

**Discussion by madmaxio** [time-rs/time#398] [Question]
Hi, I'm using this simple code

```
use time::OffsetDateTime;

fn main() {
    let now = time::OffsetDateTime::now_utc();
    let ser = serde_json::json!(now);

    println!("{:?}", ser);
    
    let de = serde_json::from_value::<OffsetDateTime>(ser);

    println!("{:?}", de);
}
```

and getting error:

```
String("2021-11-22 21:34:12.1901555 +00:00:00")

Err(Error("invalid type: string \"2021-11-22 21:34:12.1901555 +00:00:00\", expected a borrowed string", line: 0, column: 0))
```

Any hints how to fix it?

**Comment by jhpratt** [comment]
Please provide some more context. What version of `time` are you using? What features are enabled? Without basic info there's not much I can do.

**Comment by mihai-dinculescu** [comment]
I'm getting the same error with the following dependencies

```toml
serde_json = "1.0.71"
time = { version = "0.3.5", features = ["serde-human-readable"] }
```

**Comment by madmaxio** [comment]
My dependencies are same:

```
time = { version = ""0.3.5", features = ["serde-human-readable"] }
serde_json = "1.0.71"
```

**Reply by jhpratt** [reply]
Thanks. I'll look into this when I get a chance.

**Comment by madmaxio** [comment]
Perhaps any ideas for workaround? I'm using Value enum pretty heavy and need this to work.

**Reply by jhpratt** [reply]
Using threads is helpful for organization.

There is no update. I've been busy, especially with Thanksgiving. If you put together a PR I will review it when possible.

**Comment by madmaxio** [comment]
Well, I checked the place in code where deserialize is called, but no idea how to fix.

**Reply by jhpratt** [reply]
Once again, please use the existing threads. As I've said, I'll look into this when I have time. It's not a priority for me as there are a number of tests checking behavior similar to this.

**Reply by jhpratt** [reply]
After investigation, this is a limitation of `serde_json`. That crate needs to provide a way to get a `&str` when deserializing, which is something `serde` supports.

While `time` could theoretically support this by deserializing an owned `String`, I will not be doing this. Not only is it needlessly inefficient (allocating when not necessary), but it is a forward compatibility concern in that I wouldn't be able to support no-alloc human-readable parsing in the future without it being a breaking change. As such I suggest you create a feature request on the relevant repository.

**Reply by jonasbb** [reply]
This is a bug in this crate not in `serde_json`. Requesting a borrowed str while only needing a transient str (<https://serde.rs/lifetimes.html#transient-borrowed-and-owned-data>) is unfortunately a common problem. This breaks deserialization whenever an escape sequence is used or when the input cannot be borrowed from, e.g., for `Value` like types and deserialization from `Read`.

**Reply by jhpratt** [reply]
I didn't say it was a bug in `serde`, I said it was a _limitation_ of `serde_json`. Those are two very different things. Per the link that you provided, all that's needed is transient data. If you're aware of a way to do this with `serde_json::Value`, please submit a pull request.

By the way, :-1: on a response just because you disagree with it is in no way how you're going to convince me to look further into something. My time is valuable and I am providing this for free. This is a complex project and one use case failing is not a priority. No one, yourself included, has provided any information on how to solve the problem that is claimed to exist. That's fine, but don't be rude while doing so.

**Reply by jonasbb** [reply]
It is not a limitation of `serde_json`. I look into providing a PR later. The 👎 is mainly because your answer is factually wrong. All you need to get a transient string is a small `Visitor` like https://github.com/jonasbb/serde_with/blob/c427b8e32177b4547eb3b4a90647dd438a390335/serde_with/src/rust.rs#L82-L110
It also has the benefit of providing a better error message with `expecting`.
The information how to do this is in the link I provided above. I didn't mean to come across rude.

**Reply by jhpratt** [reply]
I've never used a custom visitor before, so I didn't think it'd be necessary to create one. It still seems like `serde_json` could (and imo should) provide a cleaner way to do this. Regardless I'll check this out once I finish some other work.
